social/app/models/post.py

```diff
@@ -120,3 +124,73 @@ def keys(tuple_list):
     Accepts a tuple list, returns a list of each tuple's first values
     """
     return [x[0] for x in tuple_list]
+
+
+def get_all_public_posts():
+    public_posts = dict()
+    public_posts["user_posts"] = Post.objects \
+        .filter(Q(visibility="PUBLIC") | Q(visibility="SERVERONLY")) \
+        .order_by('-published')
+    return public_posts["user_posts"]
+
+
+def get_all_friend_posts(user):
+    friend_posts = dict()
+    friend_posts["user_posts"] = Post.objects \
+        .filter(author__id__in=user.friends.all()) \
+        .filter(Q(visibility="FRIENDS") | Q(visibility="PUBLIC") | Q(visibility="SERVERONLY")) \
+        .order_by('-published')
+    return friend_posts["user_posts"]
+
+
+def get_all_foaf_posts(author):
+    friends_list = set(f.id for f in author.friends.all())
+    foafs = set()
+
+    for friend in friends_list:
+        friend_obj = Author.objects.get(pk=friend)
+        new_foafs = set(ff.id for ff in friend_obj.friends.all())
+        foafs.update(new_foafs)
+    foafs.update(friends_list)
+
+    foaf_posts = dict()
+    foaf_posts["user_posts"] = Post.objects \
+        .filter(Q(author__id__in=foafs)) \
+        .filter(Q(visibility="FOAF") | Q(visibility="PUBLIC")).order_by('-published')
+
+    return foaf_posts["user_posts"]
+
+
+# This gets all remote posts from:
+# /service/posts
+# TODO: need to query remote author to grab their friends?
+# TODO: need to save post comments too?
+def get_remote_node_posts():
+    node_posts = list()
+    for node in Node.objects.filter(local=False):
+        try:
+            for post_json in node.get_public_posts()['posts']:
+                author_json = post_json['author']
+                remote_author_id = uuid.UUID(Author.get_id_from_uri(author_json['id']))
+                Author(
+                    id=remote_author_id,
+                    node=node,
+                    displayName=author_json['displayName'],
+                ).save()
```

Comment:
I am wary about using [`save()`](https://docs.djangoproject.com/en/1.10/ref/models/instances/#saving-objects) instead of [`update_or_create()`](https://docs.djangoproject.com/en/1.10/ref/models/querysets/#update-or-create).  Please read the second paragraph of [the documentation](https://docs.djangoproject.com/en/1.10/ref/models/instances/#how-django-knows-to-update-vs-insert) that refers to a "gotcha". We'll take an extra query hit on refresh, but it's a more conservative and safer approach in this scenario.